dll/ntdll/ldr/appcompat.c

```diff
@@ -0,0 +1,386 @@
+    ASSERT(IsPowerOfTwo(VersionMask));
+
+    /* Get the VA of the name table */
+    NameTable = (PULONG)((ULONG_PTR)DllBase +
+        (ULONG_PTR)ExportDirectory->AddressOfNames);
+
+    OrdinalTable = (PUSHORT)((ULONG_PTR)DllBase +
+        (ULONG_PTR)ExportDirectory->AddressOfNameOrdinals);
+
+    /* Allocate a temp buffer for the exports */
+    Size = ExportDirectory->NumberOfNames * (sizeof(ULONG) + sizeof(USHORT));
+    OrgNameTable = RtlAllocateHeap(RtlGetProcessHeap(), 0, Size);
+    if (OrgNameTable == NULL)
+    {
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+    OrgOrdinalTable = (PUSHORT)&OrgNameTable[ExportDirectory->NumberOfNames];
+
+    /* Make a copy of the original exports */
+    RtlCopyMemory(OrgNameTable, NameTable, ExportDirectory->NumberOfNames * sizeof(ULONG));
+    RtlCopyMemory(OrgOrdinalTable, OrdinalTable, ExportDirectory->NumberOfNames * sizeof(USHORT));
+
+    /* Strip unused entries from the name and ordinal table */
+    for (i = 0, j = 0, k = 0; i < ExportDirectory->NumberOfNames; i++)
+    {
+        if (RosCompatDescriptor->ExportNameMasks[i] & VersionMask)
+        {
+            /* Put public functions into the export table */
+            NameTable[j] = OrgNameTable[i];
+            OrdinalTable[j] = OrgOrdinalTable[i];
+            j++;
+        }
+        else
+        {
+            /* Move private functions to the start of the temp buffer */
+            OrgNameTable[k] = OrgNameTable[i];
+            OrgOrdinalTable[k] = OrgOrdinalTable[i];
+            k++;
+        }
+    }
+
+    /* Finally patch the number of exports */
+    ExportDirectory->NumberOfNames = j;
+
+    /* Copy the private functions back */
+    for (i = 0; j < RosCompatDescriptor->NumberOfExportNames; i++, j++)
+    {
+        NameTable[j] = OrgNameTable[i];
+        OrdinalTable[j] = OrgOrdinalTable[i];
+    }
+
+    /* Free the copy */
+    RtlFreeHeap(RtlGetProcessHeap(), 0, OrgNameTable);
+
+    return STATUS_SUCCESS;
+}
+
+static
+NTSTATUS
+LdrpPatchExportTable(
+    _In_ PVOID DllBase,
+    _In_ PROSCOMPAT_DESCRIPTOR RosCompatDescriptor,
+    _In_ DWORD AppCompatVersion
+    )
+{
+    PIMAGE_EXPORT_DIRECTORY ExportDirectory, MagicExportDir;
+    ULONG ExportDirectorySize, OldProtect;
+    PVOID ProtectAddress;
+    SIZE_T ProtectSize;
+    PULONG FunctionTable;
+    NTSTATUS Status;
+
+    //__debugbreak();
+
+    /* Get export directory */
+    ExportDirectory = RtlImageDirectoryEntryToData(DllBase,
+                                                   TRUE,
+                                                   IMAGE_DIRECTORY_ENTRY_EXPORT,
+                                                   &ExportDirectorySize);
+    if (ExportDirectory == NULL)
+    {
+        return STATUS_INVALID_IMAGE_FORMAT;
+    }
+
+    ASSERT(RosCompatDescriptor->NumberOfExportNames == ExportDirectory->NumberOfNames);
+
+    /* Unprotect the export directory */
+    ProtectAddress = ExportDirectory;
+    ProtectSize = ExportDirectorySize;
```

Comment:
Are the tables always guaranteed to be in this section?
f.e. gcc does not put delay-loads into a separate section.